crates/ruff/src/jupyter/jupyter.rs

```diff
@@ -0,0 +1,204 @@
+}
+
+pub fn read_jupyter_notebook(path: &Path) -> Result<Option<JupyterNotebook>> {
+    let notebook: JupyterNotebook = match serde_json::from_reader(BufReader::new(File::open(path)?))
+    {
+        Ok(notebook) => notebook,
+        Err(err) => match err.classify() {
+            Category::Io => return Err(err.into()),
+            Category::Syntax | Category::Eof => {
+                // Maybe someone saved the python sources (those with the `# %%` separator)
+                // as jupyter notebook instead. Let's help them.
+                let contents = std::fs::read_to_string(path)?;
+                // Check if tokenizing was successful and the file is non-empty
+                if !(ruff_rustpython::tokenize(&contents))
+                    .last()
+                    .map_or(true, |last| last.is_err())
+                {
+                    return Err(err).context(format!(
+                        "Expected a jupyter notebook (.{} extension), \
+                        which must be internally stored as json, \
+                        but found a python source file",
+                        JUPYTER_NOTEBOOK_EXT
+                    ));
+                }
+
+                return Err(err).context(format!(
+                    "A jupyter notebook (.{}) must internally be json, \
+                    but this file isn't valid json",
+                    JUPYTER_NOTEBOOK_EXT
+                ));
+            }
+            Category::Data => {
+                // We could try to read the schema version here but if this fails it's
+                // a bug anyway
+                return Err(err).context(format!(
+                    "This file does not match the schema expected of jupyter notebooks"
+                ));
+            }
+        },
+    };
+
+    if notebook.nbformat != 4 {
+        // bail because we should have already failed at the json schema stage
+        bail!(
+            "Expected jupyter notebook format 4, found {}",
+            notebook.nbformat
+        )
+    }
+    if !notebook
+        .metadata
+        .language_info
+        .as_ref()
+        .map_or(true, |language| language.name == "python")
+    {
+        debug!(
+            "Skipping {} because it's not a python notebook",
+            path.display()
+        );
+        return Ok(None);
+    }
+
+    Ok(Some(notebook))
+}
+
+/// Concatenates all cells into a single virtual file and builds an index that maps the content
+/// to notebook cell locations
+pub fn concat_notebook(notebook: JupyterNotebook) -> (String, JupyterIndex) {
+    let mut jupyter_index = JupyterIndex {
+        // Enter a line number (1-based), get back the cell (1-based)
+        // 0 index is just padding
+        row_to_cell: vec![0],
+        // Enter a line number (1-based), get back the row number in the cell (1-based)
+        // 0 index is just padding
+        row_to_row_in_cell: vec![0],
+    };
+    let mut contents = Vec::new();
```

Comment:
Is there any way to hint the size here? (Maybe not!)

Reply:
i've also checked that `.join("\n")` does size hints internally